### PR TITLE
Workflow engine follow-ups: skill-served tools, threshold-aware consent, scheduled capabilities

### PR DESCRIPTION
**Pull request by siddseethepalli**
## Summary
Post-merge refinements to the sandboxed workflow orchestration engine (PR #34630), still behind the `workflows` feature flag (default OFF):

1. **Skill-served tools** — `run_workflow` / `manage_workflows` moved out of the always-on tool manifest into a flag-gated `workflows` bundled skill (`skill_load` + `skill_execute`), so they cost ~zero context until loaded. Adds a `manage_workflows` `list_profiles` action.
2. **Configurable leaf default** — `workflowLeaf` call-site default profile is configurable in the settings UI (already was) and now gated behind the `workflows` flag.
3. **Threshold-aware consent** — the side-effecting-workflow launch *and* HTTP-route resume approval now respects the user's auto-approve posture: at full access ("dangerously skip permissions") it no longer prompts; in normal posture it prompts exactly as before. The gate only relaxes for the explicit full-access posture.
4. **Scheduled capabilities** — scheduled workflows now carry a persisted capability manifest (`capabilities_json` on `cron_jobs`, migration 290), consented-to once at `schedule_create` (threshold-aware), used by both firing paths. Legacy/null schedules fall back to the read-only baseline.

Plus docs (`workflows.md`, `ARCHITECTURE.md`) updated to match.

## Self-review result
PASS after one remediation round (dead-code cleanup from the cutover + two stale comments; security core verified sound across two integration passes — threshold gate keys on the right variable, consent gate survives the `skill_execute` name-preservation path, `schedule_update` can't bypass creation-time consent, both firing sites baseline correctly, migration idempotent/append-only, threshold reader fails closed).

## PRs merged into this feature branch
- #34957 export tool impls
- #34958 workflows SKILL.md
- #34959 gate workflowLeaf in settings UI
- #34960 threshold-aware fresh-approval (launch)
- #34961 schedule capabilities migration (290)
- #34963 schedule-store manifest plumbing
- #34964 route resume honors full access
- #34965 skill cutover + list_profiles
- #34966 firing sites use stored manifest
- #34967 schedule_create consent
- #34968 docs

### Fix PRs (self-review remediation)
- #34969 correct stale workflow-gating comments
- #34970 remove dead Tool objects after cutover; retarget baseline drift-guard at SKILL.md

Part of plan: workflow-engine-followups.md

### CODE REVIEW

**Comment by siddseethepalli** [comment]
<details>
<summary>Plan papertrail: Workflow Engine Follow-ups</summary>

# Workflow Engine Follow-ups

## Overview

Four post-merge refinements to the sandboxed workflow orchestration engine (PR #34630, behind the `workflows` feature flag, default OFF): (1) move `run_workflow` / `manage_workflows` behind a flag-gated `workflows` bundled skill so they cost ~zero context until loaded, and teach the model to list inference profiles; (2) gate the already-configurable `workflowLeaf` call-site behind the `workflows` flag in the settings UI; (3) make the side-effecting-workflow launch/resume approval threshold-aware so a full-access ("dangerously skip permissions") posture never prompts while normal postures still do; (4) let scheduled workflows carry a persisted capability manifest consented-to once at schedule-creation time, instead of being hardcoded to the read-only baseline.

Cross-cutting: everything stays behind the `workflows` flag. Changes 3 and 4 touch security-sensitive permission/consent code. Change 4 needs a DB migration. The per-leaf `profile` option already works (`assistant/src/workflows/leaf-runner.ts`) and is unchanged.

---

## PR 1: Export workflow tool implementations for reuse

### Depends on
None

### Branch
workflow-followups/pr-1-export-tool-impls

### Title
refactor(workflows): export run_workflow / manage_workflows executor impls

### Files
- assistant/src/tools/workflows/run-workflow.ts
- assistant/src/tools/workflows/manage-workflows.ts

### Implementation steps
1. In `run-workflow.ts`, the request logic currently lives in `executeRunWorkflow(input, context)` (the `execute` of `runWorkflowTool`). Export it as a named function: change `async function executeRunWorkflow(...)` to `export async function executeRunWorkflow(...)`. Keep `runWorkflowTool.execute = executeRunWorkflow` exactly as-is. Also export the `RUN_WORKFLOW_DESCRIPTION` string constant (add `export`) so the skill build / docs can reference the canonical authoring contract.
2. In `manage-workflows.ts`, similarly export the executor (currently `executeManageWorkflows(input, context)`): add `export` to its declaration. Keep the existing `manageWorkflowsTool` definition and `execute` wiring unchanged.
3. No behavior change. The tools remain registered exactly as today; this PR only makes the impls importable so PR 3's bundled-skill executor shims can call them without duplicating logic.

### Acceptance criteria
- `bunx tsc --noEmit` passes in `assistant/`.
- `executeRunWorkflow`, `RUN_WORKFLOW_DESCRIPTION`, and `executeManageWorkflows` are exported.
- Existing tests (`run-workflow.test.ts`, `manage-workflows`-related tests) pass unchanged: `cd assistant && bun test src/tools/workflows/run-workflow.test.ts`.

---

## PR 2: Author the `workflows` bundled-skill SKILL.md (inert, flag-gated)

### Depends on
None

### Branch
workflow-followups/pr-2-workflows-skill-md

### Files
- assistant/src/config/bundled-skills/workflows/SKILL.md

### Title
feat(skills): add workflows skill instructions (authoring contract)

### Implementation steps
1. Create `assistant/src/config/bundled-skills/workflows/SKILL.md`. Model the frontmatter on a sibling bundled skill (read `assistant/src/config/bundled-skills/app-control/SKILL.md` for the exact shape). Frontmatter must include:
   - `name: workflows`
   - `description:` one line — "Author and run sandboxed multi-agent workflows that fan work out across parallel ephemeral leaf agents."
   - `metadata.emoji` and `metadata.vellum.display-name: "Workflows"`, `category: "agentLoop"` (or the closest existing category — confirm against sibling skills).
   - **`metadata.vellum.feature-flag: "workflows"`** — this reuses the existing flag so `skill_load` rejects the skill when the flag is off (precedent: `app-control` SKILL.md; gate enforced in `assistant/src/tools/skills/load.ts` ~lines 197-205).
2. Body content = the authoring contract. Port it from the canonical sources — `RUN_WORKFLOW_DESCRIPTION` in `run-workflow.ts` and `assistant/docs/workflows.md` — restructured as skill instructions. Must prominently cover the load-bearing invariants: scripts are SYNCHRONOUS (never `await`), must begin with a pure-literal `export const meta = {...}`, must `return` their result, and determinism bans (`Date.now()`/`Math.random()`/`new Date()` throw — pass seeds/timestamps via `args`). Include the host API table (`agent`/`leaf`/`parallel`/`map`/`pipeline`/`phase`/`log`/`usage`/`workflow`/`args`), leaf options (`schema`/`label`/`profile`/`persona`), and the capability-manifest consent model.
3. Document **how to list available inference profiles** before choosing a leaf `profile`. Determine the canonical model-accessible mechanism and document it: prefer an existing CLI/route (profiles live in `llm.profiles`; the backend exposes `GET config/llm/call-sites` and there may be a `vellum` profiles command — verify) and instruct her to consult it. If no model-accessible enumeration exists, note that PR 3 adds a `list_profiles` action to `manage_workflows` and document that instead (keep this section's wording aligned with whatever PR 3 implements).
4. Document run management: `manage_workflows` actions (`status`, `abort`, `list_runs`, `resume`) and the crash-recovery/resume flow (an interrupted run is resumed by `run_id`; the journal replays the completed prefix).
5. This file is inert on its own — with no `TOOLS.json` and the flag defaulting off, nothing surfaces it yet. PR 3 wires it.

### Acceptance criteria
- `assistant/src/config/bundled-skills/workflows/SKILL.md` exists with valid frontmatter matching the sibling-skill schema and `feature-flag: "workflows"`.
- The body covers: synchronous/no-await, literal `meta`, must-`return`, determinism bans, host API, leaf options, capability/consent model, profile-listing, and run management/resume.
- No code references it yet; `bunx tsc --noEmit` in `assistant/` is unaffected.

---

## PR 3: Move workflow tools into the skill (cutover) + list_profiles

### Depends on
PR 1, PR 2

### Branch
workflow-followups/pr-3-skill-cutover

### Title
feat(workflows): serve run_workflow / manage_workflows via the workflows skill

### Files
- assistant/src/config/bundled-skills/workflows/TOOLS.json
- assistant/src/config/bundled-skills/workflows/tools/run-workflow.ts
- assistant/src/config/bundled-skills/workflows/tools/manage-workflows.ts
- assistant/src/tools/tool-manifest.ts
- assistant/src/tools/registry.ts
- assistant/src/config/bundled-tool-registry.ts (regenerated)
- assistant/src/tools/workflows/manage-workflows.ts (add `list_profiles` action)
- relevant test files (see steps)

### Implementation steps
1. Create `TOOLS.json` declaring both tools, mirroring `assistant/src/config/bundled-skills/playbooks/TOOLS.json`. Each entry: `name` (`run_workflow` / `manage_workflows`), `description` (terse — the full contract lives in SKILL.md), `category`, `risk` (`low` — the executor still promotes side-effecting launches; see PR 5), `input_schema` (copy from the current tool definitions), `executor` (`tools/run-workflow.ts` / `tools/manage-workflows.ts`), and `execution_target: "host"` (runs in-process via the bundled-tool registry — confirmed pattern for bundled tools).
2. Create the two executor shims. Each exports `run(input, context)` (the `SkillToolScript` contract — see `assistant/src/tools/skills/script-contract.ts`) and delegates to the impl exported in PR 1:
   ```ts
   import { executeRunWorkflow } from "../../../../tools/workflows/run-workflow.js";
   export async function run(input, context) { return executeRunWorkflow(input, context); }
   ```
   Verify the `ToolContext` threaded through `skill_execute` → `executor.execute` → skill-tool runner carries `conversationId` and `trustClass` (the run-workflow impl calls `resolveTrustContext`/`findConversation`). If any field is dropped on the skill path, thread it through.
3. Add a `list_profiles` action to `manage_workflows` (in `assistant/src/tools/workflows/manage-workflows.ts`): return the names (and labels, if cheaply available) of profiles defined in `llm.profiles` from the resolved config, so the model can look up valid `profile` values from inside a loaded workflows skill. Keep it read-only/low-risk. (If PR 2's investigation found an existing model-accessible profile-listing path, skip this and ensure SKILL.md points at that instead — the two PRs must agree.)
4. Remove the tools from the always-on manifest: in `assistant/src/tools/tool-manifest.ts` delete the `workflowTools` array, `getWorkflowToolsIfEnabled()`, and the static `runWorkflowTool`/`manageWorkflowsTool` imports; in `assistant/src/tools/registry.ts` remove the `getWorkflowToolsIfEnabled()` registration loop (~lines 908-911). This is the atomic half of the cutover — two tools cannot share a name across both the registry and the skill, so the daemon-native registration must be removed in the same PR that activates the skill tools.
5. Regenerate the bundled-tool registry so the new in-process executors are importable in compiled binaries: `cd assistant && bun run scripts/generate-bundled-tool-registry.ts` (confirm the exact script name) and commit `assistant/src/config/bundled-tool-registry.ts`.
6. Register/discover the skill: confirm how bundled skills are catalogued (filesystem scan vs. an explicit index) by reading how `app-control`/`playbooks` are registered, and add `workflows` the same way if an explicit list exists.
7. Update tests: any test asserting `run_workflow`/`manage_workflows` are in the core tool manifest must change to assert they are NOT (they're skill tools now). Add/adjust a test that invoking `run_workflow` via the `skill_execute` path resolves to the executor and still triggers the requireFreshApproval gate for a side-effecting manifest (the executor keys on the inner tool name `run_workflow`, which `skill_execute` preserves). Run: `cd assistant && bun test src/tools/workflows/run-workflow.test.ts` and any skill-dispatch test.

### Acceptance criteria
- `run_workflow` / `manage_workflows` are absent from the always-on tool manifest; loading the `workflows` skill (flag on) makes them invocable via `skill_execute`, and they are unavailable when the flag is off.
- A side-effecting `run_workflow` invoked through `skill_execute` still hits the `requireFreshApproval` consent gate (executor sees inner name `run_workflow`).
- `manage_workflows` exposes `list_profiles` (or SKILL.md documents an existing equivalent) — verified by a test or manual run listing `llm.profiles` names.
- `bunx tsc --noEmit` passes; regenerated `bundled-tool-registry.ts` is committed; targeted tests pass.

---

## PR 4: Gate the `workflowLeaf` call-site behind the workflows flag in settings UI

### Depends on
None

### Branch
workflow-followups/pr-4-gate-workflowleaf-ui

### Title
feat(web): show the Workflow Leaf model setting only when workflows is enabled

### Files
- apps/web/src/domains/settings/ai/call-site-overrides-modal.tsx
- apps/web/src/domains/settings/ai/__tests__/ (the modal's test, if present)

### Implementation steps
1. `workflowLeaf` is already in the call-site catalog (`assistant/src/config/schemas/call-site-catalog.ts:85-91`, displayName "Workflow Leaf") and the modal auto-enumerates all catalog entries, so it already renders unconditionally. Add feature-flag gating mirroring the existing `analyzeConversation` precedent in `call-site-overrides-modal.tsx` (~lines 119-142).
2. Read the workflows flag via the existing store hook (the `workflows` flag is registered in `apps/web/src/lib/feature-flags/feature-flag-registry.json`): `const workflowsEnabled = useAssistantFeatureFlagStore.use.workflows();`.
3. Extend the `gatedCallSites` `useMemo` filter so `workflowLeaf` is excluded when `!workflowsEnabled` (compose with the existing `analyzeConversation` filter rather than replacing it).
4. Update/extend the modal test to assert `workflowLeaf` is hidden when the flag is off and shown when on. Follow the repo's apps/web test conventions (zustand v5 SSR: drive store values via `mock.module` stubs, not `setState`).

### Acceptance criteria
- With the `workflows` flag off, the Workflow Leaf entry does not appear in the call-site overrides modal; with it on, it appears and its profile override persists via the existing `PATCH /config` → `llm.callSites.workflowLeaf` path.
- `analyzeConversation` gating behavior is unchanged.
- apps/web type-check is clean for the changed file; the modal test passes.

---

## PR 5: Threshold-aware fresh-approval promotion (launch + tool-resume)

### Depends on
None

### Branch
workflow-followups/pr-5-threshold-aware-fresh-approval

### Title
feat(permissions): full-access posture skips the workflow fresh-approval prompt

### Files
- assistant/src/permissions/threshold.ts (new small shared helper) OR an existing permissions util
- assistant/src/tools/permission-checker.ts
- assistant/src/tools/permission-checker.test.ts (or the nearest existing permission-checker test)

### Implementation steps
1. Today `permission-checker.ts` (~lines 181-190) UNCONDITIONALLY promotes `allow → prompt` when `context.requireFreshApproval` is set. The conversation's resolved auto-approve threshold is already computed just above at ~line 161 (`conversationThreshold` / `riskThreshold`, from `getAutoApproveThreshold(policyContext.conversationId, policyContext.executionContext)`).
2. Add a tiny shared predicate `isFullAccessThreshold(threshold)` returning `true` when the threshold is at its maximum (`"high"` — the value that already auto-approves high-risk tools, i.e. the equivalent of "dangerously skip permissions"). Put it in a small shared module so PR 6 can reuse it. Confirm the "full access" UI/trust setting resolves to threshold `"high"`; if a distinct max sentinel exists, match on that.
3. Change the promotion at ~line 186 to skip when the posture is full access:
   ```ts
   if (
     context.requireFreshApproval &&
     result.decision === "allow" &&
     !isFullAccessThreshold(conversationThreshold)
   ) {
     result.decision = "prompt";
     result.reason = "Fresh approval required: per-invocation human review enforced";
   }
   ```
   In every non-max posture, behavior is byte-for-byte unchanged (still prompts). Only the explicit full-access posture bypasses — matching a direct side-effecting tool call, which a full-access user also runs without a prompt.
4. Leave the executor (`executor.ts`) requireFreshApproval-setting logic untouched — the gating decision now lives entirely in the permission-checker, keyed on posture. Leave the non-interactive guardian branch (~lines 268-310) and `persistentDecisionsAllowed` (~line 358) unchanged.
5. Tests (this is the security-sensitive core — cover both postures): (a) requireFreshApproval + normal threshold (`none`/`low`/`medium`) → decision promoted to `prompt`; (b) requireFreshApproval + full-access threshold (`high`) → stays `allow`, no prompt; (c) deny rules still win regardless of posture; (d) a non-requireFreshApproval allow is unaffected. Run: `cd assistant && bun test src/tools/permission-checker.test.ts`.

### Acceptance criteria
- A side-effecting workflow launch prompts in normal posture and does NOT prompt at full access.
- No change to deny handling, non-interactive guardian auto-approve, or `persistentDecisionsAllowed`.
- `bunx tsc --noEmit` passes; permission-checker tests pass for both postures.

---

## PR 6: Allow side-effecting workflow resume over HTTP at full access

### Depends on
PR 5

### Branch
workflow-followups/pr-6-route-resume-full-access

### Title
feat(workflows): route resume honors full-access posture instead of forbidding

### Files
- assistant/src/runtime/routes/workflow-routes.ts
- assistant/src/runtime/routes/workflow-routes.test.ts (or nearest existing test)

### Implementation steps
1. `handleResumeRun` currently throws `ForbiddenError` whenever `manifestGrantsSideEffects(run.capabilities)` is true, because the HTTP route has no interactive prompt channel. Make this posture-aware using the helper from PR 5: resolve the auto-approve threshold for the run's conversation (`getAutoApproveThreshold(run.conversationId ?? undefined, ...)`); if `isFullAccessThreshold(threshold)`, allow the resume to proceed (no prompt needed at full access); otherwise keep the existing `ForbiddenError` (a prompt is required but the route can't show one — the caller should resume via the assistant/`manage_workflows` path instead).
2. Keep the read-only / non-side-effecting resume path exactly as today (already allowed).
3. Tests: side-effecting resume via the route → 403 in normal posture; → proceeds in full-access posture; read-only resume → proceeds in both. Run: `cd assistant && bun test src/runtime/routes/workflow-routes.test.ts`.

### Acceptance criteria
- Route-based resume of a side-effecting run is forbidden in normal posture and allowed at full access; read-only resume is unaffected.
- `bunx tsc --noEmit` passes; route tests pass.

---

## PR 7: DB migration — add `capabilities_json` to the schedule table

### Depends on
None

### Branch
workflow-followups/pr-7-schedule-capabilities-column

### Title
feat(schedule): persist a capability manifest column on cron_jobs

### Files
- assistant/src/memory/migrations/NNN-schedule-capabilities.ts (new; next available number, currently ~287)
- assistant/src/memory/db-init.ts
- assistant/src/memory/schema/infrastructure.ts
- assistant/src/__tests__/ (a db-migration test mirroring existing schedule-column migration tests)

### Implementation steps
1. Create the migration following `assistant/src/memory/migrations/282-schedule-inference-profile.ts` (PRAGMA-guarded idempotent ALTER): add a nullable `capabilities_json TEXT` column to `cron_jobs` only if absent.
2. Register it in `assistant/src/memory/db-init.ts`: import the function and append it to the `migrationSteps` array, after the most recent schedule migration. Append-only — never reorder existing entries.
3. Add the Drizzle column to the `cronJobs`/`scheduleJobs` table definition in `assistant/src/memory/schema/infrastructure.ts`: `capabilitiesJson: text("capabilities_json")` (nullable).
4. Add a migration test asserting the column exists and is null for pre-existing rows (idempotent on re-run), mirroring existing `db-*` migration tests. Run: `cd assistant && bun test src/__tests__/<the new test>.ts`.

### Acceptance criteria
- Migration is idempotent and append-only; `cron_jobs` gains a nullable `capabilities_json` column; existing rows read back null.
- `bunx tsc --noEmit` passes; the migration test passes.

---

## PR 8: Thread the capability manifest through the schedule store

### Depends on
PR 7

### Branch
workflow-followups/pr-8-schedule-store-capabilities

### Title
feat(schedule): read/write capability manifest in the schedule store

### Files
- assistant/src/schedule/schedule-store.ts
- assistant/src/schedule/__tests__/ (schedule-store test)

### Implementation steps
1. Add `capabilities: unknown | null` to the `ScheduleJob` interface.
2. Add `capabilities?: unknown` to the `createSchedule()` and `updateSchedule()` param objects.
3. Serialize on write, mirroring the existing `workflowArgs` → `workflow_args_json` handling: store `JSON.stringify(capabilities)` into `capabilities_json` on insert and in the update setter (only when provided).
4. Parse on read in `parseJobRow`: `capabilities: parseOptionalJson(row.capabilitiesJson)` (reuse the existing optional-JSON helper used for `workflowArgs`).
5. Tests: round-trip a schedule with a manifest (`{ tools: ["file_write"], hostFunctions: [], persona: false }`) and assert create→read and update→read preserve it; a schedule created without one reads back `null`. Run the schedule-store test.

### Acceptance criteria
- `createSchedule`/`updateSchedule` persist a manifest; `parseJobRow` returns it; absent manifest reads back null.
- `bunx tsc --noEmit` passes; schedule-store tests pass.

---

## PR 9: Capture manifest consent at schedule creation

### Depends on
PR 8, PR 5

### Branch
workflow-followups/pr-9-schedule-create-consent

### Title
feat(schedule): accept + consent to a workflow capability manifest at creation

### Files
- assistant/src/tools/schedule/create.ts
- assistant/src/tools/executor.ts
- assistant/src/tools/schedule/__tests__/ (create-tool test)

### Implementation steps
1. In `assistant/src/tools/schedule/create.ts`, accept an optional `capabilities` input for workflow-mode schedules. Validate/normalize it through `CapabilityManifestSchema` + the workflow forbidden-tool checks (reuse `resolveCapabilities`/`normalizeCapabilityManifest` from `assistant/src/workflows/capabilities.ts`) so a schedule can never store an invalid or forbidden manifest. Pass the normalized manifest into both `createSchedule()` calls.
2. Consent timing: a side-effecting scheduled manifest must be approved ONCE, at creation, when a human is present. In `executor.ts`, where the workflow launch path already sets `context.requireFreshApproval` for a side-effecting manifest, extend the same recognition to a `schedule_create` (or `manage_schedules` create) call whose `mode === "workflow"` and whose `capabilities` grant side effects — set `requireFreshApproval` so creation routes through the (now threshold-aware, PR 5) approval. At full access this is silent; in normal posture it prompts once at creation. `schedule_create` is guardian-only and runs in a live conversation, so the prompt channel exists.
3. Do not prompt for read-only or absent manifests (`manifestGrantsSideEffects` is false).
4. Tests: creating a workflow schedule with a side-effecting manifest sets requireFreshApproval (prompts in normal posture, silent at full access); an invalid/forbidden manifest is rejected at creation; a read-only/absent manifest creates without a prompt; the stored manifest matches what was approved. Run the create-tool test.

### Acceptance criteria
- Workflow schedules persist a validated manifest; side-effecting ones require creation-time approval (silent at full access, one prompt otherwise); forbidden manifests are rejected; read-only ones need no prompt.
- `bunx tsc --noEmit` passes; create-tool tests pass.

---

## PR 10: Fire scheduled workflows under their stored manifest

### Depends on
PR 8

### Branch
workflow-followups/pr-10-scheduled-run-uses-manifest

### Title
feat(schedule): scheduled + run-now workflows use the persisted manifest

### Files
- assistant/src/schedule/scheduler.ts
- assistant/src/runtime/routes/schedule-routes.ts
- assistant/src/schedule/__tests__/ and/or assistant/src/runtime/routes/__tests__/ (scheduler + run-now tests)

### Implementation steps
1. In `scheduler.ts` (~lines 481-499), replace the hardcoded `manifest: { tools: [], hostFunctions: [], persona: false }` with the stored manifest: `manifest: normalizeCapabilityManifest(job.capabilities)` (which yields the read-only baseline when `job.capabilities` is null/empty — confirm `normalizeCapabilityManifest` returns the `{tools:[],hostFunctions:[],persona:false}` shape for null; if not, pass an explicit baseline fallback). Remove the "V1 LIMITATION" comment.
2. In `schedule-routes.ts` `handleRunScheduleNow` (~lines 977-993, the workflow branch), make the identical replacement. Remove the "V1 LIMITATION" comment.
3. Keep `trustContext: INTERNAL_GUARDIAN_TRUST_CONTEXT` on both paths — consent was captured at creation (PR 9), so the unattended run is authorized.
4. Tests: a schedule with a stored side-effecting manifest fires the run with that manifest; a legacy schedule (null capabilities) fires with the read-only baseline; assert on both the auto-fire (scheduler) and run-now (route) paths. Run the scheduler + route tests.

### Acceptance criteria
- Both firing paths pass the persisted manifest (baseline fallback for legacy/null); guardian trust context retained.
- `bunx tsc --noEmit` passes; scheduler + run-now tests pass.

---

## PR 11: Update workflow docs and architecture

### Depends on
PR 3, PR 5, PR 6, PR 9, PR 10

### Branch
workflow-followups/pr-11-docs

### Title
docs(workflows): skill invocation, threshold-aware approval, scheduled capabilities

### Files
- assistant/docs/workflows.md
- ARCHITECTURE.md

### Implementation steps
1. In `assistant/docs/workflows.md`: document that `run_workflow`/`manage_workflows` are served via the flag-gated `workflows` skill (loaded with `skill_load`, invoked via `skill_execute`) rather than always-on tools; document how to list inference profiles; update the approval section to state the launch/resume prompt is threshold-aware (full-access bypasses, normal posture prompts); replace the "Scheduler `workflow` mode … only the read-only baseline" paragraph with the new model (a persisted manifest consented-to once at schedule creation; legacy schedules default to read-only baseline).
2. In `ARCHITECTURE.md`, update the Workflow Orchestration Engine section to reflect the skill-based tool surface, the threshold-aware consent gate, and scheduled capability manifests. Update any Mermaid diagram that shows the tool/consent/scheduler flow.
3. Remove any now-stale references to the hardcoded read-only scheduled baseline or always-on workflow tools.

### Acceptance criteria
- `workflows.md` and `ARCHITECTURE.md` match shipped behavior (skill invocation, threshold-aware approval, scheduled capabilities); no stale V1-limitation language remains.
- Any Mermaid diagrams reflect the current flow.

</details>

**Comment by siddseethepalli** [comment]
**Review focus (risk: medium — touches permission/consent code):**

- **`assistant/src/tools/permission-checker.ts`** (PR #34960) — the one-line change gating the `requireFreshApproval` allow→prompt promotion on `!isFullAccessThreshold(riskThreshold)`. This is the security-load-bearing change: confirm full-access (`"high"`) is genuinely the user's explicit "skip permissions" posture and that no other `requireFreshApproval` caller (bundle install, dynamic skill load) is unintentionally relaxed. Other postures are byte-for-byte unchanged.
- **`assistant/src/runtime/routes/workflow-routes.ts`** (PR #34964) — `handleResumeRun` now allows side-effecting resume at full access instead of always forbidding. Note the `"conversation"` execution-context choice for the threshold lookup.
- **`assistant/src/tools/executor.ts` + `assistant/src/tools/schedule/create.ts`** (PR #34967) — creation-time consent: a side-effecting workflow-mode `schedule_create` arms `requireFreshApproval`. Confirm the manifest is validated (forbidden tools rejected) before persistence and that `schedule_update` has no path to set capabilities.
- **`normalizeCapabilityManifest(null)`** → read-only baseline at both firing sites (`scheduler.ts`, `schedule-routes.ts`, PR #34966) — backward-compat for legacy schedules.

Feature stays flag-gated OFF, so this is inert until `workflows` is enabled per-instance. CI runs on this PR (the milestone PRs were merged with --skip-ci).

File: assistant/src/schedule/scheduler.ts [hunk 1]
```diff
@@ -495,7 +492,10 @@ export async function runScheduleDueWorkOnce(
             job.wakeConversationId ??
             job.createdFromConversationId ??
             undefined,
-          manifest: { tools: [], hostFunctions: [], persona: false },
+          // The schedule's persisted capability manifest scopes the run; a
+          // legacy schedule with null `capabilities` normalizes to the read-only
+          // baseline.
+          manifest: normalizeCapabilityManifest(job.capabilities),
```
[the remark this answers is not in the document]
Comment (siddseethepalli) [reply assistant/src/schedule/scheduler.ts:498]: Thanks — I dug into this and don't think the described failure is reachable. `schedule_create` and the scheduled firing path both resolve the manifest through the **same** `resolveCapabilities()` against the global `getTool()` registry, so creation and fire-time see the identical tool set:

- Skill-provided tools (e.g. `messaging_send`) are never in the global registry — they're conversation-projected (`projectSkillTools`), and loading a skill into a conversation does not register its tools globally. So a manifest naming a skill tool is rejected **at creation** with `unknown_tool`, not silently created and failed at fire. (`create.ts` calls `resolveCapabilities(manifest)` at creation; `run-manager.start()` calls it synchronously at fire.)
- Additionally, nearly all skill tools are `execution_target: "host"`, which `resolveCapabilities` hard-rejects for any workflow leaf — a pre-existing engine guard, independent of scheduling.

So a scheduled workflow can only be created with tools that resolve in the global registry (core tools), which are present at both creation and fire.

The only fire-time-resolution residual is a *globally-registered, non-host* tool (e.g. an MCP tool) removed between creation and fire. That path fails **closed**: `start()` resolves synchronously, the scheduler's try/catch records a failed run and logs the error — no crash, no escalation — and `schedule_create` already instructs the assistant to warn when a required integration may be disconnected at runtime. The suggested fix (persist/restore skill registrations) wouldn't address that MCP case.

Resolving as not-reachable. The broader observation — workflow leaves can't use skill-provided tools at all — is a pre-existing property of the engine, tracked as a follow-up in #34972.
